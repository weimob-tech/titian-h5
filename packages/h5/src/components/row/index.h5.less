@import '../common/static/utils.less';

:host {
  display: block;
}

@row: ~'@{project}-row';

.@{row} {
  &-flex {
    display: flex;
    flex-flow: row wrap;
  }

  &::after {
    content: '';

    display: table;
    clear: both;
  }
}
