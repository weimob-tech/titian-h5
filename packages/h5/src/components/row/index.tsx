import { Component, Element, h, Prop, Watch } from '@stencil/core';
import { BasicComponentAbstract } from '../common/basic/BasicComponent';
import { addShadowRootStyle } from '../common/utils';
import { join } from '../common/utils/namespace';
import { getChildren } from '../common/utils/relation';

@Component({
  tag: 'ti-row',
  styleUrls: {
    pc: 'index.pc.less',
    h5: 'index.h5.less',
  },
  shadow: true,
})
export class TiRow implements BasicComponentAbstract {
  @Element() host: HTMLElement;

  @Prop() extCss = '';

  @Prop()
  public extStyle = {};

  @Prop() public extClass: string;

  /**
   * 列元素之间的间距
   *
   * @type number
   * @default 0
   * @example
   * <ti-row gutter="{{ 16 }}">
   *  <ti-col span="6">col 6</ti-col>
   *  <ti-col span="6">col 6</ti-col>
   *  <ti-col span="6">col 6</ti-col>
   *  <ti-col span="6">col 6</ti-col>
   * </ti-row>
   * @since 0.1.0
   */
  @Prop() public gutter = 0;

  /**
   * 是否是 flex 布局
   *
   * @type boolean
   * @default true
   * @example
   * <ti-row flex>
   *  <ti-col span="6">col 6</ti-col>
   *  <ti-col span="6">col 6</ti-col>
   *  <ti-col span="6">col 6</ti-col>
   *  <ti-col span="6">col 6</ti-col>
   * </ti-row>
   * @since 0.1.0
   */
  @Prop() public flex = true;

  get getChildren() {
    return getChildren({
      host: this.host,
      tag: 'ti-col',
      relations: 'descendant',
      useSlot: true,
    });
  }

  private completedStyle() {
    const number = this.gutter / 2;
    return {
      marginLeft: `-${number}px`,
      marginRight: `-${number}px`,
      ...this.extStyle,
    };
  }

  @Watch('gutter')
  setGutter(value, oldValue?) {
    if (value !== oldValue) {
      const children = this.getChildren;

      children?.forEach(child => {
        if (child.updateDataFromParent) {
          child.updateDataFromParent(this).then();
        }
      });
    }
  }

  componentWillLoad() {
    this.setGutter(this.gutter);

    addShadowRootStyle.call(this);
  }

  render() {
    const completedStyle = this.completedStyle();
    return (
      <div part={this.extClass} class={`${join('row', { flex: this.flex })} ${this.extClass}`} style={completedStyle}>
        <slot />
      </div>
    );
  }
}
