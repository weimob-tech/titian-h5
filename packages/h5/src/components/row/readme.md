# ti-row

<!-- Auto Generated Below -->


## Properties

| Property   | Attribute   | Description      | Type      | Default     |
| ---------- | ----------- | ---------------- | --------- | ----------- |
| `extClass` | `ext-class` | 额外的类名，添加到根节点的元素上 | `string`  | `undefined` |
| `extStyle` | --          | 额外的样式            | `{}`      | `{}`        |
| `flex`     | `flex`      | 是否是 flex 布局      | `boolean` | `true`      |
| `gutter`   | `gutter`    | 列元素之间的间距         | `number`  | `0`         |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
