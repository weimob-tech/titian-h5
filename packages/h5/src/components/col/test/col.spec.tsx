import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { TiCol } from '../index';
import { TiRow } from '../../row';

describe('渲染布局组件', () => {
  it('基本渲染 span 24', async () => {
    const page = await newSpecPage({
      components: [TiCol, TiRow],
      template: () => (
        <ti-row>
          <ti-col span={24}>span-24</ti-col>
        </ti-row>
      ),
    });
    expect(page.root).toMatchSnapshot();
  });

  it('基本渲染 span 12', async () => {
    const page = await newSpecPage({
      components: [TiCol, TiRow],
      template: () => (
        <ti-row>
          <ti-col span={12}>span-12</ti-col>
          <ti-col span={12}>span-12</ti-col>
        </ti-row>
      ),
    });
    expect(page.root).toMatchSnapshot();
  });

  it('间距渲染 span 12, gutter 16', async () => {
    const page = await newSpecPage({
      components: [TiCol, TiRow],
      template: () => (
        <ti-row gutter={16}>
          <ti-col span={12}>span-12</ti-col>
          <ti-col span={12}>span-12</ti-col>
        </ti-row>
      ),
    });
    expect(page.root).toMatchSnapshot();
  });

  it('偏移量渲染 span 12, offset 12', async () => {
    const page = await newSpecPage({
      components: [TiCol, TiRow],
      template: () => (
        <ti-row>
          <ti-col span={12} offset={12}>
            span-12
          </ti-col>
        </ti-row>
      ),
    });
    expect(page.root).toMatchSnapshot();
  });

  it('单独 col', async () => {
    const page = await newSpecPage({
      components: [TiCol, TiRow],
      template: () => <ti-col span={12}>span-12</ti-col>,
    });
    expect(page.root).toMatchSnapshot();
  });
});
