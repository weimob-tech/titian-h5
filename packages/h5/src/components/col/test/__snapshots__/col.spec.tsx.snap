// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`渲染布局组件 偏移量渲染 span 12, offset 12 1`] = `
<ti-row>
  <mock:shadow-root>
    <div class="titian-row titian-row-flex undefined" style="margin-left: -0px; margin-right: -0px;">
      <slot></slot>
    </div>
  </mock:shadow-root>
  <ti-col>
    <mock:shadow-root>
      <div class="titian-col titian-col-12 titian-col-flex-12 titian-col-offset-12 undefined" style="padding-left: 0px; padding-right: 0px;">
        <slot></slot>
      </div>
    </mock:shadow-root>
    span-12
  </ti-col>
</ti-row>
`;

exports[`渲染布局组件 单独 col 1`] = `
<ti-col>
  <mock:shadow-root>
    <div class="titian-col titian-col-12 titian-col-flex-12 titian-col-offset-0 undefined" style="padding-left: 0px; padding-right: 0px;">
      <slot></slot>
    </div>
  </mock:shadow-root>
  span-12
</ti-col>
`;

exports[`渲染布局组件 基本渲染 span 12 1`] = `
<ti-row>
  <mock:shadow-root>
    <div class="titian-row titian-row-flex undefined" style="margin-left: -0px; margin-right: -0px;">
      <slot></slot>
    </div>
  </mock:shadow-root>
  <ti-col>
    <mock:shadow-root>
      <div class="titian-col titian-col-12 titian-col-flex-12 titian-col-offset-0 undefined" style="padding-left: 0px; padding-right: 0px;">
        <slot></slot>
      </div>
    </mock:shadow-root>
    span-12
  </ti-col>
  <ti-col>
    <mock:shadow-root>
      <div class="titian-col titian-col-12 titian-col-flex-12 titian-col-offset-0 undefined" style="padding-left: 0px; padding-right: 0px;">
        <slot></slot>
      </div>
    </mock:shadow-root>
    span-12
  </ti-col>
</ti-row>
`;

exports[`渲染布局组件 基本渲染 span 24 1`] = `
<ti-row>
  <mock:shadow-root>
    <div class="titian-row titian-row-flex undefined" style="margin-left: -0px; margin-right: -0px;">
      <slot></slot>
    </div>
  </mock:shadow-root>
  <ti-col>
    <mock:shadow-root>
      <div class="titian-col titian-col-24 titian-col-flex-24 titian-col-offset-0 undefined" style="padding-left: 0px; padding-right: 0px;">
        <slot></slot>
      </div>
    </mock:shadow-root>
    span-24
  </ti-col>
</ti-row>
`;

exports[`渲染布局组件 间距渲染 span 12, gutter 16 1`] = `
<ti-row>
  <mock:shadow-root>
    <div class="titian-row titian-row-flex undefined" style="margin-left: -8px; margin-right: -8px;">
      <slot></slot>
    </div>
  </mock:shadow-root>
  <ti-col>
    <mock:shadow-root>
      <div class="titian-col titian-col-12 titian-col-flex-12 titian-col-offset-0 undefined" style="padding-left: 8px; padding-right: 8px;">
        <slot></slot>
      </div>
    </mock:shadow-root>
    span-12
  </ti-col>
  <ti-col>
    <mock:shadow-root>
      <div class="titian-col titian-col-12 titian-col-flex-12 titian-col-offset-0 undefined" style="padding-left: 8px; padding-right: 8px;">
        <slot></slot>
      </div>
    </mock:shadow-root>
    span-12
  </ti-col>
</ti-row>
`;
